core-services/src/main/java/org/zalando/nakadi/service/publishing/EventsProcessor.java

```diff
@@ -10,84 +10,183 @@
+        @Override
+        public String toString() {
+            return "Batch{et=" + eventType + ",size:" + size + "}";
+        }
+    }
+
+    private void scheduleSendBatchedRequest(final BatchedRequest req) {
+        final Runnable r = new Runnable() {
+            @Override
+            public void run() {
                 try {
-                    final JSONObject event = events.poll(pollTimeout, TimeUnit.MILLISECONDS);
-                    if (event != null) {
-                        if (jsonArray == null) {
-                            jsonArray = new JSONArray();
+                    eventPublisher.processInternal(req.data.toString(), req.eventType, false, null, false);
+                } catch (final RuntimeException ex) {
+                    LOG.info("Failed to send single batch for unknown reason", ex);
+                }
+            }
+
+            @Override
+            public String toString() {
+                return "Batch to " + req.eventType + " of size " + req.data.length();
+            }
+        };
+        executorService.submit(r);
+    }
+
+    private void dispatch() {
+        final Map<String, BatchedRequest> batchesBeingAssembled = new HashMap<>();
+        try {
+            long nextTimeCheck = System.currentTimeMillis() + batchCollectionTimeout;
+            while (true) {
+                long currentTime = System.currentTimeMillis();
+                final EventToPublish data = eventsQueue.poll(
+                        Math.max(nextTimeCheck - currentTime, 1), TimeUnit.MILLISECONDS);
+                currentTime = System.currentTimeMillis();
+                boolean batchWasSent = false;
+                if (data != null) {
+                    BatchedRequest batch = batchesBeingAssembled.get(data.eventType);
+                    if (null == batch) {
+                        batch = new BatchedRequest(data.eventType, currentTime + batchCollectionTimeout);
+                        batchesBeingAssembled.put(data.eventType, batch);
+                    }
+                    if (batch.add(data.object) >= maxBatchSize) {
+                        scheduleSendBatchedRequest(batch);
+                        batchesBeingAssembled.remove(batch.eventType);
+                        currentTime = System.currentTimeMillis();
+                        batchWasSent = true;
+                    }
+                }
+                if (batchWasSent || currentTime > nextTimeCheck) {
+                    nextTimeCheck = currentTime + batchCollectionTimeout;
+                    for (final Map.Entry<String, BatchedRequest> entry : batchesBeingAssembled.entrySet()) {
+                        if (entry.getValue().finishCollectionAt < currentTime) {
+                            scheduleSendBatchedRequest(entry.getValue());
+                            batchesBeingAssembled.remove(entry.getKey());
+                        } else if (nextTimeCheck > entry.getValue().finishCollectionAt) {
```

Comment:
i do not understand the check here. `nextTimeCheck` is aways greater than `finishCollectionAt` except the case you checked on line 162, because it is calculated as `nextTimeCheck = currentTime + batchCollectionTimeout`.

Reply:
discussed offline